src/eyreish/macros.rs

```diff
@@ -117,52 +146,113 @@ macro_rules! ensure {
             return $crate::private::Err($crate::miette!($err));
         }
     };
+    ($cond:expr, $($key:ident = $value:expr),+; $($fmt:tt)+) => {
+        if !$cond {
+            return $crate::private::Err(
+                $crate::miette!($($key = $value),+; $($fmt)+)
+            );
+        }
+    };
     ($cond:expr, $fmt:expr, $($arg:tt)*) => {
         if !$cond {
             return $crate::private::Err($crate::miette!($fmt, $($arg)*));
         }
     };
 }
 
-/// Construct an ad-hoc error from a string.
+/// Construct an ad-hoc [`Report`].
 ///
-/// This evaluates to an `Error`. It can take either just a string, or a format
-/// string with arguments. It also can take any custom type which implements
-/// `Debug` and `Display`.
+/// # Examples
 ///
-/// # Example
+/// With string literal and interpolation:
+/// ```
+/// # use miette::miette;
+/// let x = 1;
+/// let y = 2;
+/// let report = miette!("{x} + {} = {z}", y, z = x + y);
+/// assert_eq!(report.to_string().as_str(), "1 + 2 = 3");
 ///
+/// let report = miette!("{x} + {y} = {x + y}");
+/// assert_eq!(report.to_string().as_str(), "1 + 2 = 3");
 /// ```
-/// # type V = ();
-/// #
-/// use miette::{miette, Result};
 ///
-/// fn lookup(key: &str) -> Result<V> {
-///     if key.len() != 16 {
-///         return Err(miette!("key length must be 16 characters, got {:?}", key));
-///     }
+/// With [`diagnostic!`]-like arguments:
+/// ```
+/// use miette::{miette, LabeledSpan, Severity};
 ///
-///     // ...
-///     # Ok(())
-/// }
+/// let source = "(2 + 2".to_string();
+/// let report = miette!(
+///     // Those fields are optional
+///     severity = Severity::Error,
+///     code = "expected::rparen",
+///     help = "always close your parens",
+///     labels = vec![LabeledSpan::at_offset(6, "here")],
+///     url = "https://example.com"; // <- semicolon separates fields from message
+///
+///     // Rest of the arguments are passed to `format!`
+///     // to form diagnostic message
+///     "expected closing ')'"
+/// )
+/// .with_source_code(source);
 /// ```
 ///
 /// ## `anyhow`/`eyre` Users
 ///
 /// You can just replace `use`s of the `anyhow!`/`eyre!` macros with `miette!`.
 #[macro_export]
 macro_rules! miette {
-    ($msg:literal $(,)?) => {
-        // Handle $:literal as a special case to make cargo-expanded code more
-        // concise in the common case.
-        $crate::private::new_adhoc($msg)
-    };
     ($err:expr $(,)?) => ({
         use $crate::private::kind::*;
         let error = $err;
         (&error).miette_kind().new(error)
     });
-    ($fmt:expr, $($arg:tt)*) => {
-        $crate::private::new_adhoc(format!($fmt, $($arg)*))
+    ($($key:ident = $value:expr),+; $($fmt:tt)+) => {
+        $crate::Report::from(
+            $crate::diagnostic!($($key = $value),+; $($fmt)+)
+        )
+    };
+    ($($fmt:tt)+) => {
```

Comment:
First case with expression doesn't allow interpolation with single literal: 
```rs
let x = 1
let y = 2
miette!("{x} + {y}") // doesn't work as expected
miette!("{x} + {}", y) // works as expected
```

Reply:
Do you have an idea, how we can work around this?

Reply:
I feel like interpolation is more desirable than using expression in this macro, but removing it will be breaking change

Reply:
What if you used `$fmt:literal` and got rid of the semicolon thing? I think that would be good enough to disambiguate

Reply:
Done. However, integer literals support (`miette!(0)`) is now lost. Can't find a way to differentiate them with string literals.